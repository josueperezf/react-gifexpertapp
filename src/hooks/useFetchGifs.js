import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

// todo hook empiesa con la palabra hook
export const useFetchGifs = (categoria)=>{
    const [state, setstate] = useState({
        data:   [],
        loading: true
    });

    // esto se coloca es para que si cambia el componente lo que este en use useEffect no se llame constantemente
    // el useEffect se llamara la primera vez y si llegara a cambiar el valor de categoria, como no cambia no se ejectuta esa funcion useEffect de nuevo
    useEffect(() => {
        getGifs(categoria)
            .then((imgs)=>{
                setstate({
                    data:   imgs,
                    loading: false
                });
            })
    }, [categoria])

    return state; // { data:[], loading:true }
}