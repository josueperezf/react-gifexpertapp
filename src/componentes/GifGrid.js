import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';
import PropTypes from 'prop-types'

const GifGrid = ({categoria}) => {
    const {data:imagenes, loading} = useFetchGifs(categoria);
    /*
    // el siguiente hook hace evaita que cada vez que se realice un cambio en el html realice la llamada http
    useEffect (() => {
        getGifs(categoria).then((imgs)=>{
            setimages(imgs);
        });
    },[categoria])
    */
    // en la anterior linea, [categoria] se puede quitar, pero fernando la coloca y dice que es para que si en algun momento cambia el valor de la categoria, entonces todo lo que esta dentro de 'useEffect' se ejecutaria nuevamente
    
    return (
        <>
        <h3 className=" animate__animated animate__fadeIn">{categoria}</h3>
        {/** la siguiente linea es como un operador ternario, pero como no quiere evaluar el de lo contrario, se unsa el && */}
        { loading && <p className="animate__animated animate__flash" > Cargando</p> }
        <div className="card-grid">
            { /**  
             * en la seccion GifGridItem se le va a enviar el img, pero fernando hacer una destructuracion {...img}
             * con ello logra que GifGridItem no reciba asi ({img}), sino directamente ({id, title, utl})
            */}
                {
                    
                    imagenes.map(img=>{
                        return <GifGridItem key={img.id}  {...img} />
                    })
                    
                }
        </div>
        </>
    )
}

GifGrid.propTypes = {
    categoria: PropTypes.string.isRequired
}

export default GifGrid
