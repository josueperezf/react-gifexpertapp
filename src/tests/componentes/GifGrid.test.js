import { shallow } from "enzyme";
import GifGrid from "../../componentes/GifGrid";

// 
import { useFetchGifs } from "../../hooks/useFetchGifs";
// jest.mock sirve para simular que mi hook useFetchGifs ya obtuvo la data del servidor
/**
 * es donde se guardan los datos sobre cómo se llamó a la función y qué función devolvió. La .mockpropiedad también rastrea el valor de thispara cada llamada, por lo que también es posible inspeccionar esto:
*/

jest.mock('../../hooks/useFetchGifs');

describe('pruebas del componente GifGrid.js',()=>{
    const categoria = 'goku';
    const gifs = [{
        "id": "oxbNORcXx76F2",
        "title": "opm GIF",
        "url": "https://media0.giphy.com/media/oxbNORcXx76F2/giphy.gif?cid=ed0298b533ss8ce6vzekx98hbb4mlh4k8pvhahiumjlgdlje&rid=giphy.gif"
    }];

    useFetchGifs.mockReturnValue({
        data: [],
        loading: true,
    });
    const wrapper = shallow(<GifGrid categoria={categoria} />);
    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar items cuando se cargan imagenes', () => {
        // hacer que mi componente crea q ya tiene la informacion
        // useFetchGifs
        // la siguiente linea es para emular como si el useFetchGifs ya fue a la api y trajo imagenes, con el mockReturnValue simulo la data que deberia retornar el backend
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });
        const wrapper = shallow(<GifGrid categoria={categoria} />);
        expect(wrapper).toMatchSnapshot();
        // como el parrafo solo se muestra si no tiene data, entonces pregunto si existe en el html la etiqueta p, si retorna false, estamos bien
        expect(wrapper.find('p').exists()).toBe(false);
        // evaluamos si se esta dibujando en el html la etiqueta GifGridItem, decimos que la cantidad de equitas que se dibujan debe ser igual al numero de elementos que tiene el array gifs
        expect(wrapper.find('GifGridItem').length).toBe( gifs.length);
        
    });
    
    
});