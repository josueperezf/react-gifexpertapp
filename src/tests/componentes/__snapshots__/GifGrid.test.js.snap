// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente GifGrid.js debe de mostrar items cuando se cargan imagenes 1`] = `
<Fragment>
  <h3
    className=" animate__animated animate__fadeIn"
  >
    goku
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="oxbNORcXx76F2"
      key="oxbNORcXx76F2"
      title="opm GIF"
      url="https://media0.giphy.com/media/oxbNORcXx76F2/giphy.gif?cid=ed0298b533ss8ce6vzekx98hbb4mlh4k8pvhahiumjlgdlje&rid=giphy.gif"
    />
  </div>
</Fragment>
`;

exports[`pruebas del componente GifGrid.js debe mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className=" animate__animated animate__fadeIn"
  >
    goku
  </h3>
  <p
    className="animate__animated animate__flash"
  >
     Cargando
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
